posts/A-roadmap-to-TDD-adoption.md

```diff
@@ -0,0 +1,53 @@
+---
+
+You've heard all about how Test Driven Development (TDD) will solve all your problems on a software project. Your defects go away, your time to market decreases, customer satisfaction goes through the roof. Everything is supposed to be awesome if your team just did TDD. However every time you try to adopt it, the adoption fails. Why?
+
+# Test Driven Development
+
+Let's just start with what is TDD. This practice states that for any code that you write you first start writing the unit test for it. This test should fail, then you write the code that will satisfy the test, if needed you refactor and finally you write the next failing test and start this cycle again.
+
+The main reason for adoption is your IT manager or tech lead has read a blog or attended a talk on how TDD solves these main problems:
+
+* Too many defects in production
+* Long QA cycles
+* New features break old features
+* Devops adoption
+
+# Traditional approach
+
+The typical adoption pattern is to start with some classroom training on how to write unit tests, and on how to mock. You follow up with some coding katas where the team has to write a common software problem using TDD (fizz-buzz, prime numbers, bowling score calculator, etc...). Once you've got a handle on those exercises you're supposed to take that back to your project and be able to develop 'real' code using TDD. This almost always fails. First, production code is much more complicated than exercises, than code that isn't written with testing in mind is ridiculously hard to do TDD with, and finally TDD isn't actually about testing and most teams focus on writing tests for test coverage and forget the most important part of the cycle. Refactoring.
+
+TDD isn't as much about testing as it's about having testable code. If your code is easily testable, it's most likely modular and has followed good design principles. If you don't follow good design principles you'll find yourself in mocking hell where your test code becomes much more complicated than your production code, at which point you give up on unit testing and thus TDD fails. Check out [Giving up on TDD](https://blog.cleancoder.com/uncle-bob/2016/03/19/GivingUpOnTDD.html) by Robert Martin for more about how TDD and design are interlinked.
+
+# An alternative approach
+
+We have to all understand that TDD is an advanced coding practice. This is something that takes time to learn and requires you to have mastered some prerequisites.
+
+First you'll need to want to write better code, this isn't something that can be forced upon you. To start your journey as a software craftsman familiarize yourself with [SOLID design principles](https://en.wikipedia.org/wiki/SOLID), and read the Clean Code book. Design is the single most important aspect of TDD. You need to design your production code, you need to design your test code and you need to design your interfaces so that they can be seamlessly be used between production and test code.
```

Comment:
Need some kind of link to the Clean Code book (I'm assuming by Robert C. Martin), preferably the author's page.

Reply:
I didn't find one and I didn't want to link to a 3rd party like amazon